libcontainer/container_linux.go

```diff
@@ -201,18 +201,28 @@ func (c *Container) Set(config configs.Config) error {
 
 // Start starts a process inside the container. Returns error if process fails
 // to start. You can track process lifecycle with passed Process structure.
-func (c *Container) Start(process *Process) error {
+func (c *Container) Start(process *Process) (retErr error) {
 	c.m.Lock()
-	defer c.m.Unlock()
+	defer func() {
+		if retErr != nil {
+			c.terminate(process)
+		}
+		c.m.Unlock()
```

Comment:
nit: I'd keep this in a separate defer. They are semantically different things, and go defer's guarantee the order, so I'd keep them separate.

This applies to all the files, of course :)

Reply:
SGTM, I'll try it later.

Reply:
I concur; keeping the `defer c.m.Unlock()` as is helps readability a lot.